.github/CONTRIBUTING.md

```diff
@@ -66,22 +65,184 @@ To format all files, run the following script in rocBLAS directory:
+in an unnamed namespace, or be placed in the `rocblas` namespace if they need to be accessible in other compilation units. For example:
+    ```c++
+    namespace
+    {
+        // Private internal implementation
+    } // namespace
+
+    extern "C"
+    {
+        // Public C interfaces
+    } // extern "C"
+    ```
+
+7. If new arithmetic datatypes (like `rocblas_bfloat16`) are created, then unless they correspond *exactly* to a predefined system type, they should be wrapped into a `struct`, and not simply be a `typedef` to another type of the same size, so that their type is unique and can be differentiated from other types. Right now `rocblas_half` is `typedef`ed to `uint16_t`, which unfortunately prevents `rocblas_half` and `uint16_t` from being differentiable. If `rocblas_half` were simply a `struct` with a `uint16_t` member, then it would be a distinct type. It is legal to convert a pointer to a standard-layout class/struct to a pointer to its first element, and vice-versa.
+
+
+8. [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) classes should be used instead of explicit `new`/`delete`, `hipMalloc`/`hipFree`, `malloc`/`free`, etc. RAII classes are automatically exception-safe because their destructor gets called during unwinding. They only have to be declared once to construct them, and they are automatically destroyed when they go out of scope. This is better than having to match `new`/`delete` `malloc`/`free` calls in the code, especially when exceptions or early returns are possible. Even if the operation does not allocate and free memory, if it represents a change to a state which must be restored later, the it belongs in an RAII class.
+
+
+9. For code brevity and readability, when converting to numeric types, function-style casts are preferred to `static_cast<>()`. For example, `T(x)` is preferred to `static_cast<T>(x)`. When writing general containers or other classes which can accept *arbitrary* types as template parameters, not just *numeric* types, then the specific cast (`static_cast`, `const_cast`, `reinterpret_cast`) should be used, to avoid surprises. But when converting to *numeric* types, which have very well-understood behavior and are *side-effect free*, `type(x)` is more compact and clearer than `static_cast<type>(x)`. For pointers, C-style casts are okay, such as `(T*)A`.
+
+
+10. With the new device memory allocation system, rocBLAS kernels should not call `hipMalloc() ` or `hipFree()` in their own code, but should use the new device memory manager. rocBLAS kernels in the public API must also detect and respond to device memory size queries. **There is another whole
+document on this.**
+
+
+11. With some BLAS functions, especially BLAS1 functions with two vectors, the `incx` and `incy` arguments can be negative, which means the vector is treated backwards from the end. A simple trick to handle this, is to adjust the pointer to the end of the vector if the increment is negative, as in:
```

Comment:
I think you can be more specific and say:

For BLAS2 functions and BLAS1 functions with 2 vectors, the `incx` and or  `incy` arguments...